test(performance): Add 90% storage utilization scaleout/scalein test

Lakshmipathi opened the pull request:
Perform scale-out around 90% disk utilization and cluster scale-in by decommissioning the newly added nodes while measuring latency under mixed-load stress.

Testcase: https://github.com/scylladb/scylla-cluster-tests/issues/9156

### Testing
<!-- Add links to Argus/Jenkins of test test done with this PR -->
<!-- This would help the reviewer to cross check what was tested, and and review the results as needed -->
- [ ]  [waiting for completion](https://jenkins.scylladb.com/job/scylla-staging/job/LakshmipathiGanapathi/job/scylla-enterprise-perf-regression-latency-650gb-elasticity/95/)

### PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [ ] I added the relevant `backport` labels
- [ ] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

pehala commented:
Please fix all checks

pehala commented on sdcm/nemesis.py:
```diff
@@ -5625,6 +5687,13 @@ class GrowShrinkClusterNemesis(Nemesis):
     def disrupt(self):
         self.disrupt_grow_shrink_cluster()
 
+class FullStorageUtilizationNemesis(Nemesis):
```
I would appreciate more documentation for this class and all methods you added

Lakshmipathi commented on sdcm/nemesis.py:
```diff
@@ -225,6 +225,7 @@ class Nemesis:  # pylint: disable=too-many-instance-attributes,too-many-public-m
     manager_operation: bool = False  # flag that signals that the nemesis uses scylla manager
     delete_rows: bool = False  # A flag denotes a nemesis deletes partitions/rows, generating tombstones.
     zero_node_changes: bool = False
+    full_storage_utilization: bool = False
```
I added a new label to so that 90% SCT job can be grouped under it. Initially kept it as 'disrupt' but existing SCT runs may use this when `SisyphusMonkey` invoked or any job that uses `disrupt` label - which will result in job's failure. Since this job expect specific instance type and disk usage state to begin with, added it under new label.

pehala replied:
I really dont like the solution, we effectively want to exclude Nemesis from general use and to do that we create another group, which will always contain one nemesis. Isnt there a better way how to do this?

Lakshmipathi commented on sdcm/nemesis.py:
```diff
@@ -2104,11 +2105,11 @@ def _truncate_cmd_timeout_suffix(self, truncate_timeout):  # pylint: disable=no-
-    def disrupt_truncate(self):
-        keyspace_truncate = 'ks_truncate'
+    def disrupt_truncate(self, full_storage_utilization=False):
+        keyspace_truncate = 'ks_truncate' if not full_storage_utilization else 'refill_keyspace'
```
ok, update the PR.

pehala commented on performance_regression_test.py:
```diff
@@ -218,6 +218,12 @@ def _stop_load_when_nemesis_threads_end(self):
             self.loaders.kill_stress_thread()
 
+    def stop_load_during_nemesis(self):
+        with EventsSeverityChangerFilter(new_severity=Severity.NORMAL,  # killing stress creates Critical error
```
I would downgrade it to DEBUG, I dont see any value in seeing this ever, as it is expected

pehala commented:
Please use common folder structure as seen in https://github.com/scylladb/scylla-cluster-tests/pull/9796

pehala commented:
Changing to draft, as this PR needs to be adjusted to use 3 racks

mykaul commented:
If it's not too late for this, or perhaps it could be done in a follow-up, I've already discussed with @pehala a more interesting and efficient way to test this:
1. Scale out - say, to 6, or 9 nodes.
2. Fill the cluster to 40+% (in case of 6 nodes) or 25+% (in case of 9 nodes)
3. Scale *in* - you will now have >80% (in case of 6 nodes) or >75% (in case of 9 nodes) filled in already.
4. Now go all the way to 90, and then you can continue the test - additional scale-out, or remove some data and scale-in, etc.

The idea is that you'll have more 'horsepower' for ingest to fill the cluster faster.

pehala commented:
> If it's not too late for this, or perhaps it could be done in a follow-up, I've already discussed with @pehala a more interesting and efficient way to test this:
> 
>     1. Scale out - say, to 6, or 9 nodes.
> 
>     2. Fill the cluster to 40+% (in case of 6 nodes) or 25+% (in case of 9 nodes)
> 
>     3. Scale _in_ - you will now have >80% (in case of 6 nodes) or >75% (in case of 9 nodes) filled in already.
> 
>     4. Now go all the way to 90, and then you can continue the test - additional scale-out, or remove some data and scale-in, etc.
> 
> 
> The idea is that you'll have more 'horsepower' for ingest to fill the cluster faster.

Thanks for the suggestion, it is being tested in a separate effort

pehala commented:
@Lakshmipathi What is missing for this PR to be ready for merging?

Lakshmipathi commented:
> @Lakshmipathi What is missing for this PR to be ready for merging?

@pehala , I updated PR yesterday, I think its ready for final review and merge.

pehala commented:
> @pehala , I updated PR yesterday, I think its ready for final review and merge.

I see merge conflict and the actual PR is in draft

Lakshmipathi commented:
> > @pehala , I updated PR yesterday, I think its ready for final review and merge.
> 
> I see merge conflict and the actual PR is in draft

rebased the PR now.

cezarmoise commented on sdcm/nemesis.py:
```diff
@@ -2124,11 +2124,11 @@ def _truncate_cmd_timeout_suffix(self, truncate_timeout):  # pylint: disable=no-
 
-    def disrupt_truncate(self):
-        keyspace_truncate = 'ks_truncate'
+    def disrupt_truncate(self, refill_keyspace=False):
```
Why not name the keyspace in the config `ks_truncate` to avoid having to modify this?

cezarmoise commented on sdcm/nemesis.py:
```diff
@@ -4411,6 +4411,86 @@ def _shrink_cluster(self, rack=None, new_nodes: list[BaseNode] | None = None):
+        InfoEvent(message=f"Start grow cluster by {add_nodes_number} data nodes").publish()
+        new_nodes = []
+        stress_queue = None
+        for idx in range(add_nodes_number):
```
You can use `self._grow_cluster()` instead, it will also distribute nodes to racks, and it works with `parallel_node_opperations=true`

cezarmoise commented on sdcm/nemesis.py:
```diff
@@ -4411,6 +4411,86 @@ def _shrink_cluster(self, rack=None, new_nodes: list[BaseNode] | None = None):
+        time.sleep(self.interval)
+        return new_nodes
+
+    def scalein_to_reach_full_storage(self, rack=None, new_nodes: list[BaseNode] | None = None):
```
You can use `self._shrink_cluster(rack=None, new_nodes=new_nodes)` instead of a new function.

Lakshmipathi replied:
With the simulated_rack:3 , seems like I can just re-use existing GrowShrinkNemesis (because when new node added data is distributed among that rack nodes only while other rack node are still at 90%. This means we dont need run c-s too to re-fill). Will update the PR.

Lakshmipathi commented on sdcm/nemesis.py:
```diff
@@ -4411,6 +4411,86 @@ def _shrink_cluster(self, rack=None, new_nodes: list[BaseNode] | None = None):
+                new_nodes += self.add_new_nodes(count=1, rack=rack_idx,
+                                                instance_type=self.tester.params.get('nemesis_grow_shrink_instance_type'))
+                # Before starting refilling data current c-s need to be stopped
+                self.tester.stop_load_during_nemesis()
```
With my runs, keeping original mixed-workload running and starting a new write workload to re-fill data back to 90% and then adding a node lead to latency issue due to higher CPU load.  That's the reason for stopping it before starting another c-s to re-fill data.

cezarmoise commented:
With latest version, does the test no longer fill the cluster between the scale-out and scale-in?

Lakshmipathi commented:
> With latest version, does the test no longer fill the cluster between the scale-out and scale-in?

yes, refill was used to add nodes at 90% earlier. The flow used to be: 3-node cluster -> Added node-4 -> Before adding node5, since overall usage is reduced, we did refill to reach 90% again -> Added node-5 ->  decommission node4 to reach 90% again -> Drop/truncate refill keyspace -> Reaches 70% mark -> Decommission node-5 to reach 90% 

With simulated_rack:3 option, the flow now: 3-node cluster with rack1,rack2,rack3 -> added node-4 to rack1 -> rack1 data usage reduced but rack2 and rack3 disk usage still at 90% -> added node-5 to rack2 -> rack1 and rack2 data usage reduced but rack3 still at 90% -> added node-6 to rack3 -> Now overall disk usage around 50% -> If we refill data to reach 90%, then decommission will hit no space error. 

We can refill to reach 90% after scaleout but before decommission we need to drop this same data. It will be like scale-out -> refill data -> truncate refilled data -> decommission nodes.

pehala commented on test-cases/features/elasticity/perf-regression-latency-i4i_2xlarge-elasticity-90-percent.yaml:
```diff
@@ -0,0 +1,43 @@
+test_duration: 3000
```
The name of this testcase does not reflect the title or purpose of it

pehala commented:
@scylladb/qa-maintainers Please review

soyacz commented on test-cases/features/elasticity/scalein-scaleout-elasticity-90-percent.yaml:
```diff
@@ -0,0 +1,42 @@
+  "cassandra-stress write no-warmup cl=ALL n=300000000 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=1..300000000",
+  "cassandra-stress write no-warmup cl=ALL n=300000000 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=300000001..600000000",
+  "cassandra-stress write no-warmup cl=ALL n=300000000 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=600000001..900000000",
+  "cassandra-stress write no-warmup cl=ALL n=250000000 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=900000001..1200000000",
```
why 250000000?

soyacz commented on test-cases/features/elasticity/scalein-scaleout-elasticity-90-percent.yaml:
```diff
@@ -0,0 +1,42 @@
+
+simulated_racks: 3
+n_db_nodes: 3
+n_loaders: 4
```
4 loaders, but there are 5 prepare_write_cmd - for better perf, use 4 commands for prepare load

soyacz commented on test-cases/features/elasticity/scalein-scaleout-elasticity-90-percent.yaml:
```diff
@@ -0,0 +1,42 @@
+n_loaders: 4
+n_monitor_nodes: 1
+
+instance_type_loader: 'c6i.2xlarge'
```
c7i should be better.
@Lakshmipathi did you verify how much cpu is utilized on loaders? maybe we could reduce the size?

vponomaryov replied:
Loader nodes utilization:
![Screenshot from 2025-05-09 17-15-30](https://github.com/user-attachments/assets/38fb6899-f23c-4ede-91ac-3fa692e472cf)

Source: [scylla-staging/LakshmipathiGanapathi/scylla-enterprise-perf-regression-latency-650gb-elasticity#98](https://argus.scylladb.com/tests/scylla-cluster-tests/8a90107a-8c81-4332-a264-767f47351a26)

RAM-wise loader nodes have big margin, but CPU-wise it exceeds 50% utilization even without 5->4 disbalance happening in the pre-load step.

So, not changing summary load values, it may stay as-is, I think.

soyacz replied:
ok, in that case we cannot reduce this size. Unless this strange load we pose to Scylla is bogus... - @Lakshmipathi what is the reason of this higher load?

soyacz commented on test-cases/features/elasticity/scalein-scaleout-elasticity-90-percent.yaml:
```diff
@@ -0,0 +1,42 @@
+  "cassandra-stress write no-warmup cl=ALL n=300000000 -schema 'replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=200 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=1200000001..1500000000",
+]
+
+stress_cmd_w: "cassandra-stress write no-warmup cl=ALL n=335544320 -schema 'keyspace=refill_keyspace replication(strategy=NetworkTopologyStrategy,replication_factor=3)' -mode cql3 native -rate threads=300 -col 'size=FIXED(128) n=FIXED(8)' -pop seq=1500000001..1835544321"
```
is it intended to do cl=ALL? usually we base on QUORUM as this is what most users do.
btw. what is the purpose of this command? - it is not bound by duration

vponomaryov replied:
> is it intended to do cl=ALL? usually we base on QUORUM as this is what most users do. 

+1
Since we have RF=3, the disk utilization must be about the same, but CPU utilization and latency will be smaller if we use CL=Quorum.

> btw. what is the purpose of this command? - it is not bound by duration

It uses different keyspace - `refill_keyspace`, so, it adds more data/"disk utilization".

soyacz commented on test-cases/features/elasticity/scalein-scaleout-elasticity-90-percent.yaml:
```diff
@@ -0,0 +1,42 @@
+use_hdrhistogram: true
+email_recipients: ["scylla-perf-results@scylladb.com"]
+email_subject_postfix: 'fullstorage utilization test'
+parallel_node_operations: false
```
are you sure this should be false? What is the purpose of this test?

vponomaryov replied:
+1

Explicit enabling of the tablets below assume that `parallel_node_operations` should be supported.

vponomaryov commented on test-cases/features/elasticity/scalein-scaleout-elasticity-90-percent.yaml:
```diff
@@ -0,0 +1,42 @@
+test_duration: 3000
```
50 hours?

Looks too big.
Jenkins test runs may hang, so, better to reduce it

Lakshmipathi commented:
@pehala  Will update/fix this PR, next week.

fruch commented:
@pehala @Lakshmipathi is this one still needed?
